controllers/space/space_controller.go

```diff
@@ -0,0 +1,343 @@
+	// look-up the NSTemplateTier
+	tmplTier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: space.Namespace,
+		Name:      space.Spec.TierName,
+	}, tmplTier); err != nil {
+		return nil, err
+	}
+	// create the NSTemplateSet from the NSTemplateTier
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: memberOperatorNS,
+			Name:      space.Name,
+			Finalizers: []string{
+				toolchainv1alpha1.FinalizerName,
+			},
+		},
+		Spec: toolchainv1alpha1.NSTemplateSetSpec{
+			TierName: space.Spec.TierName,
+			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{
+				TemplateRef: tmplTier.Spec.ClusterResources.TemplateRef,
+			},
+			Namespaces: make([]toolchainv1alpha1.NSTemplateSetNamespace, len(tmplTier.Spec.Namespaces)),
+		},
+	}
+	for i, tmpl := range tmplTier.Spec.Namespaces {
+		nsTmplSet.Spec.Namespaces[i] = toolchainv1alpha1.NSTemplateSetNamespace{
+			TemplateRef: tmpl.TemplateRef,
+		}
+	}
+
+	return nsTmplSet, nil
+}
+
+func (r *Reconciler) ensureSpaceDeletion(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	logger.Info("terminating Space")
+	if deleted, err := r.deleteNSTemplateSet(logger, space); err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, space, r.setStatusTerminatingFailed, err, "failed to delete the NSTemplateSet")
+	} else if deleted {
+		if err := r.setStatusTerminating(space); err != nil {
+			logger.Error(err, "error updating status")
+			return false, err
+		}
+		return true, nil // requeue until NSTemplateSet is fully deleted
+	}
+	// Remove finalizer from Space
+	util.RemoveFinalizer(space, toolchainv1alpha1.FinalizerName)
+	if err := r.Client.Update(context.TODO(), space); err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, space, r.setStatusTerminatingFailed, err, "failed to remove finalizer")
+	}
+	logger.Info("removed finalizer")
+	// no need to update the status of the Space once the finalizer has been removed, since
+	// the resource will be deleted
+	return false, nil // no need to requeue
+}
+
+// deleteNSTemplateSet triggers the deletion of the NSTemplateSet on the target member cluster.
+// Returns `true/nil` if the NSTemplateSet was deleted during this call, or if it is already being deleted
+// (ie, the deletion was triggered during a previous call/reconcile loop, and it's not fully done yet)
+// Returns `false/nil` if the NSTemplateSet was already deleted (ie, resource doesn't exist anymore),
+//   or if there is no target cluster specified in the given space, of if the target cluster is unknown.
+// Returns `false/error` if an error occurred
+func (r *Reconciler) deleteNSTemplateSet(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	if space.Spec.TargetCluster == "" {
+		logger.Info("cannot deleted NSTemplateSet: no target cluster specified")
+		return false, nil // abort deletion
```

Comment:
We should clearly define what happens in the cases when the `TargetCluster` is removed from the already provisioned user. My guess would be that it would remove the NSTemplateSet from the member cluster where it was provisioned. In that case, we would need to add a `targetCluster` to the `Space.status` section.

Reply:
good point, let me add this field in the API and use it here

Reply:
this can be done in separate PR ;-)

Reply:
Tracking actual target cluster on which the NSTemplate is provisioned in faedeeb (even in case of failure)

Reply:
by saying 

>We should clearly define what happens in the cases when the TargetCluster is removed from the already provisioned user. 

I meant that we should rather create a Jira story and spend some time with it the next sprint.
For example - removing or changing the target cluster could trigger move to another member cluster - the "retarget" operation.

Reply:
well, that wasn't clear in your initial comment :/

Reply:
but ok, I will add a subtask in CRT-1308 for this particular case (which is "retargeting to another cluster", anyways)

Reply:
https://issues.redhat.com/browse/CRT-1392

Reply:
ok, but the `Space.status.targetCluster` needs an update in api repo, right?

Reply:
yes, I forgot to open the PR while dealing with the changes here: https://github.com/codeready-toolchain/api/pull/272

Reply:
this was addressed in https://github.com/codeready-toolchain/host-operator/pull/549